pym/bob/cmds/build.py

```diff
@@ -898,21 +918,110 @@ def _cookPackageStep(self, packageStep, checkoutOnly, depth):
                     BobState().setInputHashes(prettyPackagePath, [packageBuildId] + packageInputHashes)
             self._setAlreadyRun(packageStep, False, checkoutOnly)
 
+    def __queryLiveBuildId(self, step):
+        """Predict live build-id of checkout step.
+
+        Query the SCMs for their live-buildid and cache the result. Normally
+        the current result is retuned unless we're in build-only mode. Then the
+        cached result is used. Only if there is no cached entry the query is
+        performed.
+        """
+
+        key = b'\x00' + step._getSandboxVariantId()
+        if self.__buildOnly:
+            liveBId = BobState().getBuildId(key)
+            if liveBId is not None: return liveBId
+
+        if self.__verbose > 0:
+            print(colorize("   QUERY-SRC {} .. ".format(step.getPackage().getName()), "32"), end="")
+        liveBId = None
+        try:
+            liveBId = step.predictLiveBuildId()
+        finally:
+            if self.__verbose > 0:
+                if liveBId is None:
+                    print(colorize("unknown", "33"))
+                else:
+                    print(colorize("ok", "32"))
+        BobState().setBuildId(key, liveBId)
+        return liveBId
+
+    def __invalidateLiveBuildId(self, step):
+        """Invalidate last live build-id of a step."""
+
+        key = b'\x00' + step._getSandboxVariantId()
+        liveBId = BobState().getBuildId(key)
+        if liveBId is not None:
+            BobState().delBuildId(key)
+
+    def __translateLiveBuildId(self, liveBId):
+        """Translate live build-id into real build-id.
+
+        We maintain a local cache of previous translations. In case of a cache
+        miss the archive is interrogated. A valid result is cached.
+        """
+        key = b'\x01' + liveBId
+        bid = BobState().getBuildId(key)
+        if bid is not None:
+            return bid
+
+        bid = self.__archive.downloadLocalLiveBuildId(liveBId, self.__verbose)
+        if bid is not None:
+            BobState().setBuildId(key, bid)
+
+        return bid
+
+    def __getCheckoutStepBuildId(self, step, depth):
+        ret = None
+        v = (self.__verbose >= -1) and (self.__verbose < 1)
+
+        # try to use live build-ids for checkout steps
+        if not os.path.exists(step.getWorkspacePath()) and \
```

Comment:
I'm doing some tests with this and now I'm at a point where I miss my sources.
```
bob dev module_A
-> QUERY_SRC dep1
-> CHECKOUT module_A
```

I want to change something in dep1 -> no code available :(   Can you add a build argument to force the checkout?

Reply:
:rofl: I almost expected that but I've not had the chance so far to test it myself extensively. Either we add an explicit option (`--always-checkout`?) or we could resort to `--checkout-only` for the job. The drawback would be that would have to call Bob twice:

```
$ bob dev module_A
$ bob dev --checkout-only module_A/dep1
# hack...
$ bob dev module_A
```

Once the sources were fetched Bob will always use the local ones. Not sure if this is works well enough, though.

Reply:
I vote for the `--always-checkout`. Or / And a regex matching white-list for modules in my `user.yaml` where I always want to checkout the code.

Reply:
Added a `--always-checkout` option that takes a regex for packages that should be always checked out. Let's see if this works well in practice...